core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -0,0 +1,249 @@
+                    int targetShardId = OperationRouting.generateShardId(indexMetaData,
+                        Uid.decodeId(ref.bytes, ref.offset, ref.length), null);
+                    return shardId == targetShardId;
+                };
+                if (terms == null) { // this is the common case - no partitioning and no _routing values
+                    findSplitDocs(IdFieldMapper.NAME, includeInShard, leafReader, bitSet::set);
+                } else {
+                    if (indexMetaData.isRoutingPartitionedIndex()) {
+                        // this is the heaviest invariant. Here we have to visit all docs stored fields do extract _id and _routing
+                        // this this index is routing partitioned.
+                        Bits liveDocs = leafReader.getLiveDocs();
+                        Visitor visitor = new Visitor();
+                        return new ConstantScoreScorer(this, score(),
+                            new RoutingPartitionedDocIdSetIterator(leafReader, liveDocs, visitor));
+                    } else {
+                        // in the _routing case we first go and find all docs that have a routing value and mark the ones we have to delete
+                        findSplitDocs(RoutingFieldMapper.NAME, ref -> {
+                            int targetShardId = OperationRouting.generateShardId(indexMetaData, null, ref.utf8ToString());
+                            return shardId == targetShardId;
+                        }, leafReader, bitSet::set);
+                        // now if we have a mixed index where some docs have a _routing value and some don't we have to exclude the ones
+                        // with a routing value from the next iteration an delete / select based on the ID.
+                        if (terms.getDocCount() != leafReader.maxDoc()) {
+                            // this is a special case where some of the docs have no routing values this sucks but it's possible today
+                            FixedBitSet hasRoutingValue = new FixedBitSet(leafReader.maxDoc());
+                            findSplitDocs(RoutingFieldMapper.NAME, ref -> false, leafReader,
+                                    hasRoutingValue::set);
+                            findSplitDocs(IdFieldMapper.NAME, includeInShard, leafReader, docId -> {
+                                if (hasRoutingValue.get(docId) == false) {
+                                    bitSet.set(docId);
+                                }
+                            });
+                        }
+                    }
+                }
+                return new ConstantScoreScorer(this, score(), new BitSetIterator(bitSet, bitSet.length()));
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "shard_splitting_query";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return sameClassAs(o);
```

Comment:
Let's make equals/hashcode correct or throw an UOE so that we don't have caching bugs? Even if queries passed to IndexWriter.deleteDocuments are not supposed to be cached I'd like to make sure there are no issues if it occurs.